app/controllers/devise_token_auth/concerns/set_user_by_token.rb

```diff
@@ -74,6 +74,10 @@ def update_auth_header
     @client_id = nil unless @used_auth_by_token
 
     if @used_auth_by_token and not DeviseTokenAuth.change_headers_on_each_request
+      # should not append auth header if @resource related token was
+      # cleared by sign out in the meantime
+      return if @resource.reload.tokens[@client_id].nil?
```

Comment:
Yeah ... me neither. was trying to avoid that call when `change_headers_on_request` is `false`. Might need to revisit optimistic locking in general. Couple of other related issues. Merging and may revisit later!